pkg/controller/secret/controller.go

```diff
@@ -0,0 +1,360 @@
+}
+
+// getSecretValue from the referenced K8s secret.
+func (e *external) getSecretValue(ctx context.Context, cr *v1alpha1.Secret) (string, error) {
+	nn := types.NamespacedName{
+		Name:      cr.Spec.ForProvider.SecretRef.Name,
+		Namespace: cr.Spec.ForProvider.SecretRef.Namespace,
+	}
+	sc := &corev1.Secret{}
+	if err := e.kube.Get(ctx, nn, sc); err != nil {
+		return "", errors.Wrap(err, errGetK8sSecretFailed)
+	}
+
+	if cr.Spec.ForProvider.SecretRef.Key != nil {
+		val, ok := sc.Data[styraclient.StringValue(cr.Spec.ForProvider.SecretRef.Key)]
+		if !ok {
+			return "", errors.New(fmt.Sprintf(errFmtKeyNotFound, styraclient.StringValue(cr.Spec.ForProvider.SecretRef.Key)))
+		}
+		return string(val), nil
+	}
+	d := map[string]string{}
+	for k, v := range sc.Data {
+		d[k] = string(v)
+	}
+	payload, err := json.Marshal(d)
+	if err != nil {
+		return "", err
+	}
+	return string(payload), nil
+}
+
+// getChecksum from the referenced checksum secret.
+func (e *external) getChecksum(ctx context.Context, cr *v1alpha1.Secret) (string, error) {
+	if cr.Spec.ForProvider.ChecksumSecretRef == nil || cr.Spec.ForProvider.ChecksumSecretRef.Key == nil {
+		return "", errors.New(errNoChecksumRef)
+	}
+
+	nn := types.NamespacedName{
+		Name:      cr.Spec.ForProvider.ChecksumSecretRef.Name,
+		Namespace: cr.Spec.ForProvider.ChecksumSecretRef.Namespace,
+	}
+	sc := &corev1.Secret{}
+	if err := e.kube.Get(ctx, nn, sc); err != nil {
+		return "", errors.Wrap(client.IgnoreNotFound(err), errGetK8sSecretFailed)
+	}
+
+	checkSum := sc.Data[*cr.Spec.ForProvider.ChecksumSecretRef.Key]
+	return string(checkSum), nil
+}
+
+// updateChecksum in referenced checksum secret with a new value.
+func (e *external) updateChecksum(ctx context.Context, cr *v1alpha1.Secret, checksum string) error {
+	if cr.Spec.ForProvider.ChecksumSecretRef == nil || cr.Spec.ForProvider.ChecksumSecretRef.Key == nil {
+		return errors.New(errNoChecksumRef)
+	}
+
+	sc := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Spec.ForProvider.ChecksumSecretRef.Name,
+			Namespace: cr.Spec.ForProvider.ChecksumSecretRef.Namespace,
+		},
+		Data: map[string][]byte{
+			*cr.Spec.ForProvider.ChecksumSecretRef.Key: []byte(checksum),
+		},
+	}
+
+	return e.kube.Apply(ctx, sc)
+}
+
+// deleteChecksum deletes the checksum secret.
+func (e *external) deleteChecksum(ctx context.Context, cr *v1alpha1.Secret) error {
+	if cr.Spec.ForProvider.ChecksumSecretRef == nil {
+		return nil
+	}
+
+	sc := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Spec.ForProvider.ChecksumSecretRef.Name,
+			Namespace: cr.Spec.ForProvider.ChecksumSecretRef.Namespace,
+		},
+	}
+	return errors.Wrap(client.IgnoreNotFound(e.kube.Delete(ctx, sc)), errDeleteK8sSecretFailed)
+}
+
+// upsertSecret on styra
+func (e *external) upsertSecret(ctx context.Context, cr *v1alpha1.Secret, secretValue string) error {
+	req := &secrets.CreateUpdateSecretParams{
+		Context:  ctx,
+		SecretID: meta.GetExternalName(cr),
```

Comment:
`meta.GetExternalName` extracts the value from `metadata.annotations[crossplane.io/external-name]`, not `metadata.name`.

But I think you are right anyway. Secret name and ID are separated in the API. I added an extra field `name` to the spec.

Reply:
Sorry, I meant the `id` here. It would be good to have a dedicated field `spec.forProvider.id`. 

Styra makes heavy use of paths to reference secrets (e.g. `alice/repo/passphrase`: https://docs.styra.com/policies/git-management/mount-git-repos-for-libraries#mount-a-git-repository-through-the-api). It would be good to be able to create those secrets with `provider-styra`

Reply:
After offline discussion we agreed that the `crossplane.io/external-name` annotation should be used for setting the `id`. We should document this in `spec.forProvider`.